docs/ffi/ffi.md

```diff
@@ -0,0 +1,467 @@
+
+## FFI calls on a GPU
+
+So far, we have been interfacing only with foreign functions running on the CPU, but JAX's FFI also supports calls to GPU code.
+Since this documentation page is automatically generated on a machine without access to a GPU, we can't execute any GPU-specific examples here, but we will go over the key points.
+
+When defining our FFI wrapper for CPU, the function signature that we used was:
+
+```c++
+ffi::Error RmsNormImpl(float eps, ffi::Buffer<ffi::DataType::F32> x,
+                       ffi::Result<ffi::Buffer<ffi::DataType::F32>> y)
+```
+
+To update this to interface with a CUDA kernel, this signature becomes:
+
+```c++
+ffi::Error RmsNormImpl(cudaStream_t stream, float eps,
+                       ffi::Buffer<ffi::DataType::F32> x,
+                       ffi::Result<ffi::Buffer<ffi::DataType::F32>> y)
+```
+
+And the handler definition is updated to include a `Ctx` in its binding:
+
+```c++
+XLA_FFI_DEFINE_HANDLER(
+    RmsNorm, RmsNormImpl,
+    ffi::Ffi::Bind()
+        .Ctx<ffi::PlatformStream<cudaStream_t>>()
+        .Attr<float>("eps")
+        .Arg<ffi::Buffer<ffi::DataType::F32>>(/* x */)
+        .Ret<ffi::Buffer<ffi::DataType::F32>>(/* y */));
+```
+
+Then, the `RmsNormImpl` can use the CUDA stream to launch CUDA kernels.
+
+On the front end, the registration code would be updated to specify the appropriate platform:
+
+```python
+xla_client.register_custom_call_target(
+  "rms_norm_cuda", rms_norm_lib_cuda.rms_norm(), platform="CUDA", api_version=1
+)
+```
+
+### Supporting multiple platforms
+
+To support running our `rms_norm` function on both GPU and CPU, we can combine our implementation above with the {func}`jax.lax.platform_dependent` function:
+
+```{code-cell} ipython3
+def rms_norm_cross_platform(x, eps=1e-5):
+  assert x.dtype == jnp.float32
+  out_type = jax.ShapeDtypeStruct(x.shape, x.dtype)
+
+  def impl(target_name):
+    return lambda x: jex.ffi.ffi_call(
+      target_name,
+      out_type,
+      x,
+      eps=np.float32(eps),
+      vectorized=True,
+    )
+
+  return jax.lax.platform_dependent(x, cpu=impl("rms_norm"), cuda=impl("rms_norm_cuda"))
+
+
+np.testing.assert_allclose(rms_norm_cross_platform(x), rms_norm_ref(x), rtol=1e-5)
+```
+
+This version of the function will call the appropriate FFI target depending on the runtime platform.
+
+As an aside, it may be interesting to note that while the jaxpr and lowered HLO both contain a reference to both FFI targets:
+
+```{code-cell} ipython3
+jax.make_jaxpr(rms_norm_cross_platform)(x)
+```
+
+```{code-cell} ipython3
+print(jax.jit(rms_norm_cross_platform).lower(x).as_text().strip())
+```
+
+by the time the function is compiled, the appropriate FFI has been selected:
```

Comment:
You can show the `as_text(dialect="hlo")`, which shows the HLO passed to the compiler.

Reply:
I didn't know that! Updated.